### PR TITLE
Add firewall section to the networking page

### PR DESCRIPTION
**Pull request by larskarlitski**
First look at a what a firewall module might look like.

It operates on firewalld's default zone and there are no plans to expose zones, as they don't seem to be used very often on servers.

The delete buttons don't work yet, but the UI updates itself when adding and removing services from the default zone with one of:
```
sudo firewallctl zone '' remove service ssh
sudo firewallctl zone '' add service ssh
``` 

firewall

This will implement the minimal version of the design and iterate in later pull requests.

- [x] show the firewall section on the default page when firewalld is installed
- [x] show the list of services that are currently active
- [x] allow to remove a service
- [x] allow to add one of firewalld's predefined services
- [x] tests
- [x] add doc/guide page
- [ ] feature video for release notes

### CODE REVIEW

**Comment by garrett** [comment]
Related links:

- design page: https://github.com/cockpit-project/cockpit/wiki/Feature:-Firewall
- issue: https://github.com/cockpit-project/cockpit/issues/1094

**Comment by q2dg** [comment]
Ei, there's any still-alpha firewall module available to download for trying it? Thanks!

**Comment by larskarlitski** [comment]
First update towards @garrett's design: move the page over to the network page.

Small bug (in addition to all the stuff missing): `cockpit.location.go()` in the breadcrumb doesn't do what it should.

**Comment by larskarlitski** [comment]
Quick update: make the breadcrumb work correctly, hook up the switch to starting/stopping `firewalld.service`, and hide the whole section if firewalld is not installed.

**Comment by larskarlitski** [comment]
Update with a refactor, because my initial approach of talking to firewalld turned out to make too many D-Bus calls and cause too many re-renders. Removing services is now supported. Also lots of tiny fixes to get it close to the design.

File: pkg/networkmanager/firewall.jsx [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+            );
+        }
+
+        var services = [...this.state.enabledServices].map(id => this.state.services[id]);
+        services.sort((a, b) => a.name - b.name);
+
+        return (
+            <div className="container-fluid page-ct">
+                <ol className="breadcrumb">
+                    <li><a onClick={go_up}>{_("Network")}</a></li>
```
Comment (andreasn) [inline pkg/networkmanager/firewall.jsx]: Just a flyby comment that this is called "Networking" in the menu and on the breadcrumb on the other sub-pages.

Comment (larskarlitski) [reply pkg/networkmanager/firewall.jsx]: Indeed. Thanks for the catch! It'll be fixed with the next push.

**Comment by larskarlitski** [comment]
Update with minor fixes pointed out by @martinpitt and @andreasn. Also add a first version of the "Add Services" dialog. It's still missing the search box and some layout fixes to qualify for this PR, so leaving WIP.

Depends on #8794 now.

**Comment by larskarlitski** [comment]
Fix a couple of things we discussed on IRC: 

* "N Active Rules" on the main networking page is now clickable (and blue on hover)
* services on the firewall page and in the "Add Services" dialog are now (actually) sorted alphabetically
* clicking anywhere on a row in the "Add Services" dialog toggles its checkbox
* fix alignment and spacing issues

**Comment by larskarlitski** [comment]
Another update, which I think brings it to a state for a first round of review (removed WIP):
* add search box in the Add Services Dialog
* skip the tests on some images that don't have firewalld
* fix networking test that used ambiguous selector now, because of the new elements

File: pkg/networkmanager/firewall.jsx [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+                       columns={columns}
+                       tabRenderers={tabs}/>;
+}
+
+class AddServicesDialogBody extends React.Component {
+    constructor() {
+        super();
+
+        this.state = {
+            selected: new Set()
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Please initialize `services` here (to `null` so that your spinner still works). At least I prefer the state components to be spelled out here in full, for an easier mental model.

File: pkg/networkmanager/firewall.jsx [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        firewall.getAvailableServices()
+            .then(services => this.setState({ services: services }));
+    }
+
+    onToggleService(event) {
+        var service = event.target.id;
+        var enabled = event.target.checked;
+
+        this.setState(oldState => {
+            var selected = new Set(oldState.selected);
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Whee, functional `setState()` - I've learned something new today :+1: 

Please use `let` in arrow functions for correct scoping.

File: pkg/networkmanager/firewall.jsx [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+export class Firewall extends React.Component {
+    constructor() {
+        super();
+
+        this.state = {
+            installed: firewall.installed,
+            enabled: firewall.enabled,
+            services: firewall.services,
+            enabledServices: firewall.enabledServices
+        };
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: It feels redundant to keep a verbatim copy of the global `firewall` state here and in `onFirewallChanged`. Isn't it enough to just call `setState({ firewall })` in `onFirewallChanged`, and initialize it the same way here? Or am I missing something?

File: pkg/networkmanager/firewall.jsx [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+    onSwitchChanged(value) {
+        if (value)
+            firewall.enable();
+        else
+            firewall.disable();
+    }
+
+    onAddServices() {
+        var services = [...this.state.services].map(id => this.state.services[id]);
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Why does this need a copy of state.services? The .map already creates a new object.

File: pkg/networkmanager/interfaces.js [hunk 6]
```diff
@@ -2144,6 +2145,11 @@ PageNetworkInterface.prototype = {
             cockpit.location.go('/');
         });
 
+        $("#networking-firewall-link, #networking-firewall-summary").on("click", function() {
+            cockpit.jump("/network/firewall");
```
Comment (martinpitt) [inline pkg/networkmanager/interfaces.js]: Indeed it is not necessary any more to specify the host in master, but in the interest of backportability, and for working with older cockpit-system in RHEL, please add it here.

File: pkg/networkmanager/firewall.jsx [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+            if (!event || event.button !== 0)
+                return;
+
+            cockpit.jump("/network", cockpit.transport.host);
+        }
+
+        if (!this.state.installed) {
+            return (
+                <EmptyState title={_("Firewall is not available")} icon="fa fa-exclamation-circle">
+                    <p translatable="yes">Please install the <tt>firewalld</tt> package.</p>
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Can you please update the .pot and make sure that this actually works? I thought the `translate` (aka `translatable`) tags were only considered in .html files. If it doesn't, use `cockpit.format(_("Please ... {0} package"), "firewalld")` here.

File: pkg/networkmanager/firewall.jsx [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+                </EmptyState>
+            );
+        }
+
+        var addServiceAction = <button className="btn btn-primary pull-right" onClick={this.onAddServices}>{_("Add Services…")}</button>;
+
+        var services = [...this.state.enabledServices].map(id => this.state.services[id]);
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Again, why does enabledServices need to be copied?

File: test/verify/check-networking-firewall [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+from testlib import *
+import time
+
+
+def wait_unit_state(machine, unit, state, timeout=10):
+    start = time.time()
+    while machine.execute("systemctl is-active {} || true".format(unit)).strip() != state:
+        time.sleep(0.1)
+        if time.time() - start > timeout:
+            raise Exception("timed out waiting for {} to become {}".format(unit, state))
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: Consider using testlib.py's `wait()` which has the timeout and graceful exception handling etc. builit in. Something like

    wait(lambda: machine.execute("systemctl is-active " + unit).strip() != state, delay=0.2)

(with the default of `tries=60` you then get a 12s timeout)

File: test/verify/check-networking-firewall [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+
+class TestNetworking(MachineCase):
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: Please rename to `TestFirewall`, to avoid confusion.

These tests need to be blacklisted on at least rhel-7-4 (where networking is in base, and thus don't get this feature). There are also some images (like the Debians) which don't have firewalld, I suggest the test skips them for now.

File: test/verify/check-networking-firewall [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+        wait_unit_state(m, "firewalld", "inactive")
+        b.wait_in_text("#networking-firewall-summary", "0 Active Rules")
+
+        b.click("#networking-firewall-switch label:not(.active)")
+        b.wait_present("#networking-firewall-switch label.active:not(.disabled)")
+        b.wait_in_text("#networking-firewall-switch label.active", "On")
+        wait_unit_state(m, "firewalld", "active")
+
+        active_rules = m.execute("firewallctl zone '' list services").split()
+        b.wait_in_text("#networking-firewall-summary", "{} Active Rules".format(len(active_rules)))
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: Corner case: Use "Rule" here instead of "Rules", for the corner case that there is really just one. However, our images should have at least ssh and cockpit, so I guess it's fine.

File: test/verify/check-networking-firewall [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        b.wait_in_text("#networking-firewall-switch label.active", "Off")
+        m.execute("systemctl start firewalld")
+        b.wait_in_text("#networking-firewall-switch label.active", "On")
+
+        # to toggle the switch, click on the non-active label, where the slider is
+        b.click("#networking-firewall-switch label:not(.active)")
+        b.wait_present("#networking-firewall-switch label.active:not(.disabled)")
+        b.wait_in_text("#networking-firewall-switch label.active", "Off")
+        wait_unit_state(m, "firewalld", "inactive")
+        b.wait_in_text("#networking-firewall-summary", "0 Active Rules")
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: This behaviour feels a bit odd. If the firewall is disabled, then "0 active rules" feels like "enabled, but unconfigured". Shouldn't this be hidden completely, as the firewall state is already shown as "off"?

File: test/verify/check-networking-firewall [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        wait_unit_state(m, "firewalld", "active")
+
+        active_rules = m.execute("firewallctl zone '' list services").split()
+        b.wait_in_text("#networking-firewall-summary", "{} Active Rules".format(len(active_rules)))
+
+        b.click("#networking-firewall-link")
+        b.enter_page("/network/firewall")
+
+        b.click(".breadcrumb li:first a")
+        b.enter_page("/network")
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: I suggest stopping and removing firewalld.service from disk (don't forget systemctl daemon-reload), and verify the expected behaviour of "not installed".

File: test/verify/check-networking-firewall [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+
+    def testFirewallPage(self):
+        b = self.browser
+        m = self.machine
+
+        self.login_and_go("/network/firewall")
+
+        # Add a service that's probably not enabled by default anywhere
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: ... so do the equivalent wait_not_present()?

File: pkg/networkmanager/firewall-client.es6 [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+    enabledServices: new Set()
+};
+
+cockpit.event_target(firewall);
+
+var firewalld_service = service.proxy('firewalld');
+var firewalld_dbus = cockpit.dbus('org.fedoraproject.FirewallD1', { superuser: 'try' });
+
+firewalld_service.addEventListener('changed', function () {
+    var installed = !!firewalld_service.exists;
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/firewall-client.es6]: It's set to `null` before the first update. I've been bitten by JavaScript's weird type coercion too many times, so I've left the `!!`.

File: pkg/networkmanager/firewall-client.es6 [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+                return info;
+            });
+    }))
+
+    /*
+     * Work around `cockpit.all()` returning results in individual arguments -
+     * that's just confusing.
+     */
+    .then(function () {
+        return Array.prototype.slice.call(arguments);
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/firewall-client.es6]: Yeah, it's the most succinct way to do so. I can't use a fat-arrow function here, because that doesn't have the `arguments` object (took me a while to figure out). This looks confusing, because the real culprit here is `cockpit.all()` not returning a single list. I've updated the comment to be more clear.

File: pkg/networkmanager/firewall-client.es6 [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+                var info = {
+                    id: service,
+                    name: name,
+                    description: description,
+                    ports: ports.map(p => ({ port: p[0], protocol: p[1] }))
+                };
+
+                firewall.services[service] = info;
+                return info;
+            });
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/firewall-client.es6]: I'm handling those errors where the surrounding function is called.

File: pkg/networkmanager/firewall-client.es6 [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+});
+
+firewalld_dbus.subscribe({
+    interface: 'org.fedoraproject.FirewallD1.zone',
+    path: '/org/fedoraproject/FirewallD1',
+    member: 'ServiceRemoved'
+}, (path, iface, signal, args) => {
+    var [ , service, ] = args;
+
+    firewall.enabledServices.delete(service);
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/firewall-client.es6]: `firewall.services` needs to contain all services, because it is uses in the "Add Services" dialog. It's used as a cache if you never open the dialog so that we don't need to fetch information on every state update.

File: pkg/networkmanager/firewall-client.es6 [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+                               'addService', ['', service, 0])
+        .then(reply => getDefaultZonePath())
+        .then(path => firewalld_dbus.call(path, 'org.fedoraproject.FirewallD1.config.zone',
+                                         'addService', [service]));
+};
+
+firewall.addServices = function (services) {
+    return cockpit.all(services.map(s => firewall.addService(s)))
+        .then(function () {
+            return Array.prototype.slice.call(arguments);
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/firewall-client.es6]: Yeah, they return promises that resolve when the service is added. I've documented all those functions now so that that is more clear.

File: pkg/networkmanager/firewall.jsx [hunk 20]
```diff
@@ -0,0 +1,265 @@
...
+import { Listing, ListingRow } from "cockpit-components-listing.jsx";
+import { OnOffSwitch } from "cockpit-components-onoff.jsx";
+import { show_modal_dialog } from "cockpit-components-dialog.jsx";
+
+import "table.css";
+import "./networking.css";
+
+var _ = cockpit.gettext;
+
+require("page.css");
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/firewall.jsx]: Good catch! I think I used it for *something* at some point...

File: pkg/networkmanager/firewall.jsx [hunk 21]
```diff
@@ -0,0 +1,265 @@
...
+
+function EmptyState(props) {
+    return (
+        <div className="curtains-ct blank-slate-pf">
+            {props.icon && <div className="blank-slate-pf-icon"><div className={props.icon}></div></div>}
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/firewall.jsx]: Ugh, how did this happen?

File: pkg/networkmanager/firewall.jsx [hunk 22]
```diff
@@ -0,0 +1,265 @@
...
+    );
+}
+
+function ServiceRow(props) {
+    if (!props.service.name)
+        return <ListingRow key={props.service.id}
+                           columns={["", "", ""]} />
+
+    var tcp = props.service.ports.filter(p => p.protocol.toUpperCase() == 'TCP');
+    var udp = props.service.ports.filter(p => p.protocol.toUpperCase() == 'UDP');
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/firewall.jsx]: Yes, but only very few services make use of that. According to the docs:

> The protocol can be any protocol supported by the system. Please have a look at /etc/protocols for supported protocols.

File: pkg/networkmanager/firewall.jsx [hunk 23]
```diff
@@ -0,0 +1,265 @@
...
+
+    var tabs = [
+        { name: _("Details"), renderer: () => <p>{props.service.description}</p> }
+    ];
+
+    return <ListingRow key={props.service.id}
+                       rowId={props.service.id}
+                       columns={columns}
+                       tabRenderers={tabs}/>;
+}
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/firewall.jsx]: I usually do this when I write components in `pkg/lib`, i.e. when they're supposed to be used in different files from where they are defined. So I haven't added them here either. Let me know if you feel stronger about it than I do.

File: pkg/networkmanager/firewall.jsx (hunk 5 above)
Comment (larskarlitski) [reply pkg/networkmanager/firewall.jsx]: I agree, this is very unfortunate. `this.state.services` is a `Set`, which doesn't have `map` defined.

File: test/verify/check-networking-firewall (hunk 11 above)
Comment (larskarlitski) [reply test/verify/check-networking-firewall]: Wow, awesome catch!

File: test/verify/check-networking-firewall (hunk 12 above)
Comment (larskarlitski) [reply test/verify/check-networking-firewall]: This is as designed. I agree that we might want something better here. I'll ask @garrett.

**Comment by larskarlitski** [comment]
Thanks very much for the detailed review. Lots of good stuff!

To not spam you too much, I haven't commented "Fixed" on the suggestions that I accepted exactly as you wrote them.

The one thing I haven't fixed is showing the UI read-only when we're not superuser, as that's a bit more involved. How do we usually find this out? Ask polkit directly?

**Comment by martinpitt** [comment]
@larskarlitski : We spoke about this in IRC, but for the record: You can query https://www.freedesktop.org/software/polkit/docs/latest/eggdbus-interface-org.freedesktop.PolicyKit1.Authority.html#eggdbus-method-org.freedesktop.PolicyKit1.Authority.CheckAuthorization to see if your process is allowed to authorize. Of course you still need to know whether the user ticked the "use my password for admin tasks", I hope we store that somewhere?

File: pkg/networkmanager/firewall-client.es6 [hunk 24]
```diff
@@ -120,6 +121,12 @@ firewalld_dbus.subscribe({
     firewall.dispatchEvent('changed');
 });
 
+cockpit.spawn(['sh', '-c', 'pkcheck --action-id org.fedoraproject.FirewallD1.all --process $$ -u 2>&1'])
```
Comment (martinpitt) [inline pkg/networkmanager/firewall-client.es6]: A bit ugly indeed, but I can't think of a better way. If we encounter this more often in the future, we might instead query/store  the bridge pid as a global variable somewhere.

File: pkg/networkmanager/firewall.jsx [hunk 25]
```diff
@@ -270,7 +272,12 @@ export class Firewall extends React.Component {
             );
         }
 
-        var addServiceAction = <button className="btn btn-primary pull-right" onClick={this.onAddServices}>{_("Add Services…")}</button>;
+        var addServiceAction;
+        var classes = "btn btn-primary pull-right";
+        if (this.state.firewall.readonly)
+            addServiceAction = <button className={classes} disabled title={_("You are not authorized to modify the firewall.")}>{_("Add Services…")}</button>;
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: This is using the standard browser tooltips (white on black) which are "wrong". I suggest using `update_privileged()` for that, or if that somehow is impractical, use `.tooltip('fixTitle')` (but that requires jQuery) or our `Tooltip` react component.

File: pkg/networkmanager/firewall.jsx [hunk 26]
```diff
@@ -65,6 +65,8 @@ function ServiceRow(props) {
             { udp.map(p => p.port).join(', ') }
         </div>,
         <button className="btn btn-danger pficon pficon-delete"
+                title={props.readonly ? _("You are not authorized to modify the firewall.") : null}
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Also the wrong tooltip. Here it seems mostly irrelevant as an unprivileged user cannot even see the current firewall rules. But there is the theoretical possibility that an admin changes the polkit rules to allow that to users, but not modification.

File: pkg/networkmanager/firewall-client.es6 (hunk 24 above)
Comment (larskarlitski) [reply pkg/networkmanager/firewall-client.es6]: I agree. If we start doing this in more places, we probably also want (internal) library functions for that.

File: pkg/networkmanager/firewall.jsx (hunk 25 above)
Comment (larskarlitski) [reply pkg/networkmanager/firewall.jsx]: `update_privileged()` also depends on jQuery, which I don't want to pull into `firewall.jsx`. I've used the React component, which unfortunately needed some minor tweaking. I've added the two commits on here as well.

File: pkg/networkmanager/firewall.jsx (hunk 26 above)
Comment (larskarlitski) [reply pkg/networkmanager/firewall.jsx]: I can see the firewall rules as an unprivileged user on fedora. Fixed as above.

**Comment by larskarlitski** [comment]
> This works well enough, thanks for that! The "On/Off" button on the /network/firewall page still needs the same "disable" treatment as "Add services..". I can't actually operate it in unprivileged mode - it stays at "off" even though the firewall is really on.

As discussed on IRC: I can operate the button even in unprivileged mode. I guess there's something wrong with `superuser: "try"`, but that's unrelated to this pull request.

> I don't see the firewall section on /network at all in unprivileged mode. I think this is right - the /firewall page is next to useless then anyway.

I do, but I've left it the way it is for the reason above.

I've added two more commits to fix our `Tooltip` component, but left the extra commit for the unprivileged stuff so that it's easier for you to review. I'll squash it into the main commit after the review.

**Comment by martinpitt** [comment]
> I can operate the button even in unprivileged mode

Actually now I see that too, not sure what was different. However, does that actually work for you? What happens here is that the firewall is shown as "off", then I can click it to "on" (but does not actually start firewalld.service), and then it moves to the disabled state. From then on it stays disabled.

Another issue is that in privileged mode the On/Off switch is a bit confusing. It takes about two seconds after clicking to actually take effect, but during that time the button stays enabled and does not change state. This feels like "I mis-clicked", and when clicking again it interferes with the first click.

Compare this with the "Automatic Updates" on/off switch on software updates. This immediately switches state, and gets disabled until the action takes place. This was the result of some involved discussions with our designers, so I think we should have the same behaviour here.

Both are minor issues, so I would be happy if these become follow-ups if you prefer. The rest looks good, great job with figuring out the tooltip issues! I confirm that the changes work well, also on the kdump/playground pages.

**Comment by larskarlitski** [comment]
I've changed the switch on the firewall page to behave in the same way as the other one: it is disabled while the state change is pending.

The tests had two issues: one was a logic mixup (`!=` vs `==`) and the other one that we need to wait for the Add Services button to become enabled before we can click it. The latter happened only after I added the read-only mode, which is the default and the button can be disabled for a few ms.

Also folded the read-only mode commit into the main one.

Sorry for the delay.

**Comment by larskarlitski** [comment]
Still some test failures.

**Comment by martinpitt** [comment]
Hmm, this collided with #9027. Maybe this was too aggressive and we should revert the "indent" stuff..

**Comment by martinpitt** [comment]
@larskarlitski : Force-pushed with running `npm run eslint:fix` to get the indentation right, and with capitalizing the commit messages.

**Comment by larskarlitski** [comment]
Ah, I rebased and fixed a conflict,  it didn’t build again, because I had to run. Thanks!

File: pkg/networkmanager/firewall.jsx (hunk 23 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/networkmanager/firewall.jsx]: It's fine for me, just wanted to point it out.

File: pkg/networkmanager/firewall.jsx (hunk 4 above)
Comment (martinpitt) [reply pkg/networkmanager/firewall.jsx]: I'd still like to discuss the above.

File: pkg/networkmanager/firewall-client.es6 [hunk 27]
```diff
@@ -0,0 +1,198 @@
...
+    path: '/org/fedoraproject/FirewallD1',
+    member: 'ServiceRemoved'
+}, (path, iface, signal, args) => {
+    const service = args[1];
+
+    firewall.enabledServices.delete(service);
+    firewall.dispatchEvent('changed');
+});
+
+cockpit.spawn(['sh', '-c', 'pkcheck --action-id org.fedoraproject.FirewallD1.all --process $$ -u 2>&1'])
```
Comment (apollo13) [inline pkg/networkmanager/firewall-client.es6]: Maybe replace `-u` with `--allow-user-interaction` -- short options are nice for users to type; but not really in code imo.

**Comment by martinpitt** [comment]
eslint spotted a bug and has another complaint:
```
/build/cockpit/pkg/networkmanager/firewall.jsx
  179:36  error  Unknown property 'for' found, use 'htmlFor' instead  react/no-unknown-property
  253:64  error  Arrow function should not return assignment          no-return-assign
```

File: test/verify/check-networking-firewall [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+from testlib import *
+import time
+
+
+def wait_unit_state(machine, unit, state):
+
+    def active_state(unit):
+        # don't use `systemctl is-active` here because of https://bugzilla.redhat.com/show_bug.cgi?id=1073481
+        # outputs "ActiveState=state\n"
+        line = machine.execute("systemctl show -p ActiveState {}".format(unit))
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: Note: `--value` does not yet exist in current CentOS 7.4. Once that moves to 7.5, we can use it.

File: test/verify/check-networking-firewall [hunk 29]
```diff
@@ -0,0 +1,117 @@
...
+        return line.strip().split("=")[1]
+
+    wait(lambda: active_state(unit) == state, delay=0.2)
+
+
+@skipImage("firewalld not installed", "rhel-7-5", "centos-7", "fedora-atomic", "rhel-atomic", "debian-stable", "debian-testing", "ubuntu-stable", "ubuntu-1604")
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: Follow-up: We should definitively install it on rhel-7-5 and centos-7 at least. So on rhel-8 it's already present apparently, good!

File: doc/guide/feature-firewall.xml [hunk 30]
```diff
@@ -0,0 +1,27 @@
...
+
+  <para>Firewalld controls access to its APIs via PolicyKit. The user logged
+    into Cockpit needs to have the appropriate permissions to view or modify
+    the settings.</para>
+
+  <para>Cockpit can currently only show, add, and remove predefined firewalld
+    services in the default zone.</para>
+
+  <para>To perform similar tasks from the command line, use
+    <ulink url="http://www.firewalld.org/documentation/man-pages/firewall-cmd.html">firewalld-cmd</ulink>:
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply doc/guide/feature-firewall.xml]: I had tried `https://firewalld.org` and that didn't work. It does with the `www` though. Good catch!

File: doc/guide/feature-firewall.xml [hunk 31]
```diff
@@ -0,0 +1,27 @@
...
+    </para>
+
+<programlisting>
+$ <command>sudo firewall-cmd --list-services</command>
+dhcpv6-client samba-client mdns ssh cockpit
+</programlisting>
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply doc/guide/feature-firewall.xml]: Yeah good point.

File: pkg/networkmanager/firewall-client.es6 [hunk 32]
```diff
@@ -0,0 +1,198 @@
...
+                });
+    }))
+
+    /*
+     * Work around `cockpit.all()` returning results in individual arguments -
+     * that's just confusing and doesn't work with ES6 style functions.
+     */
+            .then(function () {
+                return Array.prototype.slice.call(arguments);
+            });
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/firewall-client.es6]: There's a comment above these lines which explains why this is: `cockpit.all` returns results as individual arguments, so we need access to the `arguments` object, which doesn't exist in arrow functions.

Also, eslint gets the indentation wrong here. Do you know why?

File: pkg/networkmanager/firewall-client.es6 [hunk 33]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+firewalld_dbus.addEventListener('owner', (event, owner) => {
+    firewall.enabled = !!owner;
+
+    if (!firewall.enabled) {
+        firewall.services = {};
+        firewall.enabledServices = new Set();
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/firewall-client.es6]: Indeed! Very nice catch, thanks.

File: pkg/networkmanager/firewall.jsx [hunk 34]
```diff
@@ -0,0 +1,342 @@
...
+            filter: ""
+        };
+
+        this.onFilterChanged = this.onFilterChanged.bind(this);
+        this.onToggleService = this.onToggleService.bind(this);
+    }
+
+    componentDidMount() {
+        firewall.getAvailableServices()
+                .then(services => this.setState({ services: services }));
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/firewall.jsx]: Hm, this is a normal `setState()`, not functional?!

File: pkg/networkmanager/firewall.jsx [hunk 35]
```diff
@@ -0,0 +1,342 @@
...
+        this.setState({ pendingTarget: value });
+
+        if (value)
+            firewall.enable();
+        else
+            firewall.disable();
+    }
+
+    onAddServices() {
+        var services = [...this.state.firewall.services].map(id => this.state.firewall.services[id]);
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/firewall.jsx]: You did, but no worries ;) `map()` only exists on Arrays, not Objects.

File: pkg/networkmanager/firewall.jsx [hunk 36]
```diff
@@ -0,0 +1,342 @@
...
+            if (!event || event.button !== 0)
+                return;
+
+            cockpit.jump("/network", cockpit.transport.host);
+        }
+
+        if (!this.state.firewall.installed) {
+            return (
+                <EmptyState title={_("Firewall is not available")} icon="fa fa-exclamation-circle">
+                    <p>{cockpit.format(_("Please install the {0} package"), "firewalld")}</p>
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/firewall.jsx]: Indeed!

File: pkg/networkmanager/interfaces.js [hunk 37]
```diff
@@ -2231,6 +2237,33 @@ PageNetworkInterface.prototype = {
...
+            if (!firewall.installed) {
+                $('#networking-firewall').hide();
+                return;
+            }
+
+            $('#networking-firewall').show();
+            $('#networking-firewall-switch').onoff('value', firewall.enabled);
+
+            var n = firewall.enabledServices.size;
+            var summary = cockpit.ngettext('1 Active Rule', n + ' Active Rules', n);
```
[the remark this answers is not in the document]
Comment (larskarlitski) [reply pkg/networkmanager/interfaces.js]: Oooooooops

File: test/verify/check-networking-firewall (hunk 28 above)
Comment (larskarlitski) [reply test/verify/check-networking-firewall]: Ah I didn't know about that. Thank you!

**Comment by larskarlitski** [comment]
Thanks very much for the review! The ones I've left uncommented are fixed as well. Didn't want to spam you too much ;)

File: pkg/networkmanager/firewall.jsx (hunk 34 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/networkmanager/firewall.jsx]: Err yes, clearly I grossly misread that.. :-/

File: pkg/networkmanager/firewall-client.es6 (hunk 32 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/networkmanager/firewall-client.es6]: I don't, I'm afraid. You mean the alignment of the `});`, right? If some rule gets it wrong, feel free to disable it for the time being.

**Comment by martinpitt** [comment]
Oops, the [test still fails](https://fedorapeople.org/groups/cockpit/logs/pull-8032-20180426-192020-a48e7317-verify-fedora-28/log.html#170): The [screenshot](https://fedorapeople.org/groups/cockpit/logs/pull-8032-20180426-193434-a48e7317-verify-fedora-i386/TestFirewall-testNetworkingPage-fedora-i386-127.0.0.2-2601-FAIL.png) shows that it's literally displayed `$0 Active Rules`.

File: pkg/networkmanager/interfaces.js [hunk 38]
```diff
@@ -2231,6 +2237,33 @@ PageNetworkInterface.prototype = {
...
+            if (!firewall.installed) {
+                $('#networking-firewall').hide();
+                return;
+            }
+
+            $('#networking-firewall').show();
+            $('#networking-firewall-switch').onoff('value', firewall.enabled);
+
+            var n = firewall.enabledServices.size;
+            var summary = cockpit.ngettext('$0 Active Rule', '$0 Active Rules', n);
```
Comment (martinpitt) [inline pkg/networkmanager/interfaces.js]: Sorry, this needs to be wrapped in `cockpit.format(cockpit.ngettext(...), n)` to actually resolve the `$0`.

**Comment by larskarlitski** [comment]
Added `cockpit.format()` (and a hack, because that cannot handle `0` as an argument) and reordered the location with the weird indent to make that more readable.

**Comment by martinpitt** [comment]
Corresponding semaphore run: https://semaphoreci.com/cockpit/cockpit/branches/pull-request-8032/builds/20 (wasn't added as test here for some reason)

**Comment by openface** [comment]
Just coming into this project and am really impressed.  In particular, am glad to see support for firewalld. 

Are there any plans to allow specific firewall configuration per network interface?
